### ISSUE TITLE
Update database indexes to improve search

### ISSUE THREAD

**Issue by rija** [gigascience/gigadb-website#910]
Existing database index are not optimal for the search function and and the SQL audit has revealed that index are missing.
So update existing indexes and add new indexes to avoid sequential scan of tables during search queries.

This is a follow-up to #838

**Comment by rija** [comment]
Optimisation for first (out of three) query:

```
create materialized view file_finder as
SELECT "f"."id" as id, "f"."dataset_id" as dataset_id, "fs"."sample_id" as sample_id, d.upload_status as upload_status, coalesce(f.name,'') || coalesce(f.description,'') || coalesce(a.attribute_name,'') || coalesce(fa.value,'')  as document
	FROM "file" "f"
	left join file_sample fs on f.id = fs.file_id
	left join dataset d on d.id = f.dataset_id
	left join file_attributes fa on f.id = fa.file_id
	left join attribute a on a.id = fa.attribute_id;

create index file_finder_search_idx on file_finder using GIN (to_tsvector('english',document));

explain analyse SELECT f.id, f.dataset_id, f.sample_id
	FROM file_finder f
	WHERE  to_tsvector('english',f.document) @@ to_tsquery('genome')
	AND f.upload_status = 'Published';
	

SELECT f.id, f.dataset_id, f.sample_id, f.upload_status
	FROM file_finder f
	WHERE  to_tsvector('english',f.document) @@ to_tsquery('genome & plant')
	AND f.upload_status = 'Published';
```

**Comment by rija** [comment]
Optimisation for the second query: 
```
create materialized view sample_finder as
	select s.id as id, ds.dataset_id as dataset_id, d.upload_status as upload_status, coalesce(s.name,'') || coalesce(s.consent_document, '') || coalesce(s.contact_author_name, '') || coalesce(sp.common_name, '') || coalesce(sp.genbank_name,'') || coalesce(sp.scientific_name, '') || coalesce(a.attribute_name,'') || coalesce(sa.value,'') as document 
	from sample s
		left join dataset_sample ds on ds.sample_id = s.id
		left join species sp on sp.id = s.species_id
		left join dataset d on d.id = ds.dataset_id
		left join sample_attribute sa on sa.sample_id = s.id
		left join attribute a on sa.attribute_id = a.id
		
create index sample_finder_search_idx on sample_finder using GIN (to_tsvector('english',document));

explain analyse SELECT s.id, s.dataset_id 
	FROM sample_finder s
	WHERE  to_tsvector('english',s.document) @@ to_tsquery('Plant')
	AND s.upload_status = 'Published';
	
SELECT s.id, s.dataset_id 
	FROM sample_finder s
	WHERE  to_tsvector('english',s.document) @@ to_tsquery('planting')
	AND s.upload_status = 'Published';	
```

**Comment by rija** [comment]
Optimisation for the third and last query: 

```
create materialized view dataset_finder as
	with dataset_author_fullnames as (
		select dataset_id, string_agg(a.surname || ' '||a.first_name||' ' || a.middle_name, ';') as names 
		from dataset_author da, author a 
		where da.author_id = a.id 
		group by dataset_id
	)

	, dataset_author_initialednames as (
		select dataset_id, string_agg(a.surname||', '||substring(a.first_name,1,1)||', '||substring(a.middle_name,1,1), ';') as names 
		from dataset_author da, author a 
		where da.author_id = a.id 
		group by dataset_id
	)

	, dataset_keywords as (
		select dataset_id, string_agg(value, ';') as keywords
		from dataset_attributes 
		where attribute_id=455 
		group by dataset_id
	)

	, dataset_projects as (
		select dataset_id, string_agg(p.name, ',') as names 
		from dataset_project dp, project p 
		where dp.project_id = p.id 
		group by dataset_id
	)
	
	, dataset_types as (
		select dt.dataset_id, string_agg(t.name,';') as types  
		from type t, dataset_type dt 
		where dt.type_id=t.id
		group by dt.dataset_id
	)
	
	select d.id as id, 
		d.upload_status as upload_status, 
		d.publication_date as publication_date, 
		coalesce(d.identifier,'') || coalesce(d.title,'') || coalesce(daf.names,'') || coalesce(dai.names,'') || coalesce(d.description,'') || coalesce(dt.types,'') || coalesce(dk.keywords,'') || coalesce(dp.names,'') || coalesce(m.identifier,'') || coalesce(m.pmid::varchar,'') || coalesce(df.grant_award, '') || coalesce(df.comments,'') || coalesce(fn.primary_name_display,'') as document
	from dataset d
	left join manuscript m on d.id = m.dataset_id
	left join dataset_funder df on df.dataset_id = d.id
	left join funder_name fn on fn.id = df.funder_id
	left join dataset_author_fullnames daf on daf.dataset_id = d.id
	left join dataset_author_initialednames dai on dai.dataset_id = d.id
	left join dataset_keywords dk on dk.dataset_id = d.id
	left join dataset_projects dp on dp.dataset_id = d.id
	left join dataset_types dt on dt.dataset_id = d.id

create index dataset_finder_search_idx on dataset_finder using GIN (to_tsvector('english',document));

explain analyse SELECT d.id, d.document, d.upload_status 
	FROM dataset_finder d
	WHERE  to_tsvector('english',d.document) @@ to_tsquery('penguin')
	AND d.upload_status = 'Published';

```

**Comment by rija** [comment]
TODOs on this task:

* [ ] Create Yii database migrations for the materialised views and GIN indexes above 
* [ ] Replace the queries in ``protected/components/DatabaseSearch.php`` with the three ``SELECT`` queries above
* [ ] Add ``after_save`` hook to ActiveRecord classes to refresh the materialised views upon update
* [ ] Make any ajdustement to the views search boxes if necessary

**Comment by rija** [comment]
Apparently, after the results is returned, the view templates make futher query to get additional data to present:
```
$file->location
$file->name
$file->type->name
$file->getSizeWithFormat()
$sample->dataset->shortUrl
$sample->name
$sample->species->common_name
$sample->species->tax_id
$sample->dataset->identifier
$dataset->description
$dataset->authorNames
$dataset->identifier
```

**Comment by rija** [comment]
Also those two functions unecessary make a query to the database as the same result can be achieved by filtering the resultset with a dataset id.:
```
        $dsamples = $dataset->getSamplesInIds($samples['data']);
        $dfiles = $dataset->getFilesInIds($files['data']); 
```

**Comment by rija** [comment]
implemented in PR #951
